core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1637,6 +1639,77 @@ public void sync(Translog.Location location) {
         }
     }
 
+    private final ReentrantLock fsyncLock = new ReentrantLock(); // lock order --> fsyncLock --> monitor
+    private final Object syncConsumerLock = new Object(); // guards access to the two below
+    private List<Tuple<Translog.Location, Consumer<Exception>>> syncConsumerList = new ArrayList<>();
+    private List<Tuple<Translog.Location, Consumer<Exception>>> syncConsumerListSpare = new ArrayList<>();
+
+    /**
+     * Syncs the given location with the underlying storage unless already synced. This method might return immediately without
+     * actually fsyncing the location until the sync listener is called. Yet, unless there is already another thread fsyncing
+     * the transaction log the caller thread will be hijacked to run the fsync for all pending fsync operations.
+     * This method allows indexing threads to continue indexing without blocking on fsync calls. We ensure that there is only
+     * one thread blocking on the sync an all others can continue indexing.
+     */
+    public final void sync(Translog.Location location, Consumer<Exception> syncListener) {
+        verifyNotClosed();
+        // we first add the location and the listener to the list and then try to accquire the
+        // fsync lock. if we get it we run fsync for the further most location such that we only fsync once
+        // for all pending operations.
+        synchronized (syncConsumerLock) {
+            // TODO should we make this bounded and block the caller thread?
+            // I assume this won't grow to anything crazy but it might be good to have an upperbound here to prevent OOM?
+            syncConsumerList.add(new Tuple<>(location, syncListener));
+        }
+        while (fsyncLock.tryLock()) {
+            try {
+                List<Tuple<Translog.Location, Consumer<Exception>>> candidates;
+                synchronized (syncConsumerLock) {
+                    candidates = syncConsumerList;
+                    syncConsumerList = syncConsumerListSpare;
+
+                    syncConsumerListSpare = candidates;
+                    if (syncConsumerListSpare.size() > 1024) {
+                        // we limit the size to 1024 elements to ensure we don't hold on to too large lists from peak times
+                        syncConsumerListSpare = new ArrayList<>();
+                    }
```

Comment:
what would be the downside of doing `syncConsumerListSpare = new ArrayList<>();` all the time?